### PR TITLE
ENT-4108 | added management command `populate_job_names`

### PR DESCRIPTION
**Pull request by HammadAhmadWaqas**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
- [ ] [Version](https://github.com/edx/taxonomy-connector/blob/master/taxonomy/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/taxonomy-connector/blob/master/CHANGELOG.rst) record added

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/taxonomy-connector/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/taxonomy-connector/actions), verify version has been pushed to [PyPI](https://pypi.org/project/taxonomy-connector/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [course-discovery](https://github.com/edx/course-discovery) to upgrade dependencies (including taxonomy-connector)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in course-discovery will look for the latest version in PyPi.

### CODE REVIEW

File: taxonomy/emsi_client.py [hunk 1]
```diff
@@ -177,6 +177,32 @@ def __init__(self):
         """
         super(EMSIJobsApiClient, self).__init__(scope='postings:us')
 
+    @JwtEMSIApiClient.refresh_token
+    def get_details(self, ranking_facet, query_filter):
+        """
+        Query the EMSI API for the loopup of the pre-defined filter_query.
+
+        Arguments:
+            ranking_facet (RankingFacet): Data will be fetch for this facet.
```
Comment (sameenfatima78) [inline taxonomy/emsi_client.py:186]: Nit: fetch -> fetched

File: taxonomy/emsi_client.py [hunk 2]
```diff
@@ -177,6 +177,32 @@ def __init__(self):
...
+            query_filter (dict): Filters to be sent in the POST data.
+
+        Returns:
+            dict: A dictionary containing details of all the details.
```
Comment (sameenfatima78) [inline taxonomy/emsi_client.py:190]: Can we change "details" to something more specific? Talking about the second "detail"

File: taxonomy/management/commands/populate_job_names.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from taxonomy.utils import chunked_queryset
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    command for populating missing job names
+
+    While creating Job object we only populate its id. We need to get all the Job in the system that do no have names.
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:23]: Nit: no -> not, Job -> jobs

File: taxonomy/management/commands/populate_job_names.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from taxonomy.exceptions import TaxonomyAPIError
+from taxonomy.models import Job
+from taxonomy.utils import chunked_queryset
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    command for populating missing job names
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:21]: Nit: Command*

File: taxonomy/management/commands/populate_job_names.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    While creating Job object we only populate its id. We need to get all the Job in the system that do no have names.
+
+    Example usage:
+        $ # Update the existing job names.
+        $ ./manage.py populate_job_names
+        """
+    help = 'populating missing Job names in the system.'
+
+    def _update_job(self, job_bucket):
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:31]: I think we already have a helping method with a similar name which may add confusion, how about we make this one more specific and change it to `_update_job_name` instead?

File: taxonomy/management/commands/populate_job_names.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        Persist the jobs data in the database.
+        """
+        job_id = job_bucket['id']
+        job_name = job_bucket['properties']['singular_name']
+        Job.objects.update_or_create(external_id=job_id, defaults={'name': job_name})
+
+    def _update_jobs(self):
+        """
+        Updated the Jobs by fetching data from EMSI
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:41]: Nit: 
"Updated the Jobs by fetching data from EMSI" 
->
"Fetch the Jobs with missing names from EMSI and update their names."

File: taxonomy/management/commands/populate_job_names.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                for bucket in buckets:
+                    self._update_job(bucket)
+        except TaxonomyAPIError as error:
+            message = 'Taxonomy API Error for updating the jobs for Ranking Facet {} Error: {}'.format(
+                ranking_facet, error
+            )
+            LOGGER.error(message)
+            raise CommandError(message)
+        except KeyError as error:
+            message = f'Missing keys in update Job names. Error: {error}'
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:63]: Can we use the same formatting standard everywhere to keep things consistent? Either use `f strings` or `format strings`.

File: taxonomy/management/commands/populate_job_names.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    command for populating missing job names
+
+    While creating Job object we only populate its id. We need to get all the Job in the system that do no have names.
+
+    Example usage:
+        $ # Update the existing job names.
+        $ ./manage.py populate_job_names
+        """
+    help = 'populating missing Job names in the system.'
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:29]: Nit: Populates*

File: taxonomy/management/commands/populate_job_names.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        except KeyError as error:
+            message = f'Missing keys in update Job names. Error: {error}'
+            LOGGER.error(message)
+            raise CommandError(message)
+
+    def handle(self, *args, **options):
+        """
+        Entry point for management command execution.
+        """
+        self._update_jobs()
```
Comment (sameenfatima78) [inline taxonomy/management/commands/populate_job_names.py:71]: Would be nice if we can add some logging here when this command runs successfully

File: tests/test_emsi_client.py [hunk 10]
```diff
@@ -245,3 +246,34 @@ def test_get_job_postings_data_error(self):
...
             self.client.get_job_postings(RankingFacet.TITLE_NAME, JOB_POSTINGS_FILTER)
+
+    @mock_api_response(
+        method=responses.POST,
+        url=EMSIJobsApiClient.API_BASE_URL + '/taxonomies/{}/lookup'.format(RankingFacet.TITLE.value),
+        json=JOB_LOOKUP,
+    )
+    def test_get_details(self):
+        """
+        Validate that the behavior of client while fetching job lookup data.
```
Comment (sameenfatima78) [inline tests/test_emsi_client.py:257]: Nit: Remove "that"

File: tests/test_emsi_client.py [hunk 11]
```diff
@@ -245,3 +246,34 @@ def test_get_job_postings_data_error(self):
...
+
+    @mock_api_response(
+        method=responses.POST,
+        url=EMSIJobsApiClient.API_BASE_URL + '/taxonomies/{}/lookup'.format(RankingFacet.TITLE.value),
+        json=JOB_LOOKUP,
+        status=400,
+    )
+    def test_get_details_error(self):
+        """
+        Validate that the behavior of client when error occurs while fetching job lookup data.
```
Comment (sameenfatima78) [inline tests/test_emsi_client.py:271]: Nit: Remove "that"

**Review by sameenfatima78 — APPROVED** [review]
Some modifications needed, otherwise LGTM

File: tests/test_emsi_client.py [hunk 12]
```diff
@@ -245,3 +246,34 @@ def test_get_job_postings_data_error(self):
...
+        """
+        with raises(
+                TaxonomyAPIError,
```
Comment (sameenfatima78) [inline tests/test_emsi_client.py:274]: Indentation seems to be a bit out of place here

File: taxonomy/management/commands/populate_job_names.py (hunk 9 above)
Comment (HammadAhmadWaqas) [reply taxonomy/management/commands/populate_job_names.py:71]: added

File: taxonomy/management/commands/populate_job_names.py (hunk 7 above)
Comment (HammadAhmadWaqas) [reply taxonomy/management/commands/populate_job_names.py:62]: fixed the other, now using f strings

**Review by sameenfatima78 — APPROVED** [review]
LGTM (once build passes)

File: tests/test_emsi_client.py (hunk 12 above)
Comment (HammadAhmadWaqas) [reply tests/test_emsi_client.py:274]: This is expected Indentation here. Otherwise, we get a quality error.

Comment (HammadAhmadWaqas) [reply tests/test_emsi_client.py:274]: (bad-continuation)

**Review by saleem-latif — APPROVED** [review]
Please use single-quoted strings e.g. `'test string "quoted"'` in all places, once this is done :+1:

File: taxonomy/emsi_client.py [hunk 13]
```diff
@@ -177,6 +177,32 @@ def __init__(self):
         """
         super(EMSIJobsApiClient, self).__init__(scope='postings:us')
 
+    @JwtEMSIApiClient.refresh_token
+    def get_details(self, ranking_facet, query_filter):
+        """
+        Query the EMSI API for the loopup of the pre-defined filter_query.
```
Comment (muhammad-ammar) [inline taxonomy/emsi_client.py:183]: typo: loopup -> lookup

**Review by muhammad-ammar — APPROVED** [review]

**Pull request merged**